src/Git/Index/Ignore.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Git.Index.Ignore where
+
+import Control.Applicative ((<|>))
+import Control.Monad (replicateM_)
+import Data.Attoparsec.ByteString.Char8
+  ( Parser, anyChar, char, eitherP, many', many1, peekChar, satisfy, sepBy1
+  , takeTill, sepBy', inClass)
+import qualified Data.ByteString.Char8 as Char8
+import Data.Monoid ((<>))
+import Data.Foldable ()
+import qualified System.Path as Path
+import System.Path.PartClass (AbsRel)
+
+import Git.Internal (char_, parses, SwitchFileDir)
+
+import Git.Index.Glob
+  ( DirectoryIgnorePattern(..), DoubleStar(..), Glob(..), GlobPart(..)
+  , specialChars, matchGlobPath)
+
+newtype Ignores = Ignores [IgnorePattern] deriving Monoid
+
+ignoresP :: Parser Ignores
+ignoresP = Ignores <$> ipP `sepBy'` (many1 $ satisfy $ inClass "\r\n")
+
+data DoIgnore
+  = DiIgnore
+  | DiInclude
+  | DiUnspecified
```

Comment:
I'm not sure about this constructor - in the end, you want to be able to get a definitive `Ignore` or `Include` from the logic, and you have to translate an `Unspecified` into your default. It feels to me like the pattern matching logic gives you a `Maybe DoIgnore`, which you then want to handle the combination of.
```haskell
> import  Data.Monoid as M
> doIgnoreResults = [Just DiIgnore, Nothing, Just DiInclude]
> M.getLast $ foldMap M.Last doIgnoreResults
Just DiInclude
```
That behaviour feels not quite right either because you never want the Maybe out the end. Perhaps you could even
```haskell
> data DoIgnore = DiIgnore | DiInclude
> instance Monoid DoIgnore where {mempty = DiInclude; mappend _ di = di}
> doIgnoreResults = [Just Ignore, Nothing, Just Include]
> mconcat $ mapMaybe id [Just DiIgnore, Nothing, Just DiInclude]
DiInclude
```